libbpf-rs/src/map.rs

```diff
@@ -61,15 +61,11 @@ impl OpenMap {
         unsafe { libbpf_sys::bpf_map__set_inner_map_fd(self.ptr, inner.fd()) };
     }
 
-    /// Reuse an already-pinned map for `self`.
-    pub fn reuse_pinned_map<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
-        let cstring = util::path_to_cstring(path)?;
-
-        let fd = unsafe { libbpf_sys::bpf_obj_get(cstring.as_ptr()) };
-        if fd < 0 {
-            return Err(Error::System(errno::errno()));
-        }
-
+    /**
+    * Reuse a map created within the same userspace process.
+    * To use maps across userspace processes, pin the map first and use "reuse_pinned_map".
+    */
+    pub fn reuse_fd(&self, fd: i32) -> Result<()> {
         let ret = unsafe { libbpf_sys::bpf_map__reuse_fd(self.ptr, fd) };
 
         // Always close `fd` regardless of if `bpf_map__reuse_fd` succeeded or failed
```

Comment:
I don't think we should be closing the file descriptor here. In the case of `reuse_pinned_map`, it made more sense to close the fd since we were opening the fd first with `bpf_obj_get`.

Reply:
Makes sense, since the ``fd`` in such a scenario should be closed by first owner or when it gets dropped. Nice catch @insearchoflosttime .
@thesw4rm thoughts/reasons otherwise?

Reply:
> I don't think we should be closing the file descriptor (on line 78) for the standalone `reuse_fd`. In the case of `reuse_pinned_map`, it made more sense to close the fd since we opened the fd first with `bpf_obj_get`.

Good point. I'll change the line to only close it within `reuse_pinned_map`. Why are we able to close the fd for the pinned map and still use that fd within the BPF program?

Reply:
> Why are we able to close the fd for the pinned map and still use that fd within the BPF program?

For the pinned map case, the fd isn't exposed outside the `reuse_pinned_map` function whereas in the non-pinned map case, the fd is able to be used outside the function.

Reply:
> > Why are we able to close the fd for the pinned map and still use that fd within the BPF program?
> 
> For the pinned map case, the fd isn't exposed outside the `reuse_pinned_map` function whereas in the non-pinned map case, the fd is able to be used outside the function.

Then how exactly does the userspace/BPF program use the pinned map if the fd is already closed/not available outside the function? Does userspace open a new stream to the map after reuse_pinned_map is called?

Reply:
libbpf's bpf_map__reuse_fd() is doing dup3() internally and creates a new FD out of the one provided. So it's ok to close the original FD that was opened from BPF FS. When reusing map by FD directly, we don't own/control original FD, so we shouldn't close it.

Reply:
That makes sense thank you!